src/persist/src/indexed/mod.rs

```diff
@@ -541,14 +541,41 @@ impl<U: Buffer, L: Blob> Indexed<U, L> {
             .traces
             .get_mut(&id)
             .ok_or_else(|| Error::from(format!("never registered: {:?}", id)))?;
-        let since = Antichain::from_elem(since);
+        let prev = trace.since();
+        let seal = trace.get_seal();
 
-        trace.allow_compaction(since)?;
+        if seal.less_equal(&since) {
+            return Err(Error::from(format!(
+                "invalid compaction at or in advance of trace seal {:?}: {:?}",
+                seal, since,
+            )));
+        }
+
+        if !prev.less_than(&since) {
+            return Err(Error::from(format!(
+                "invalid compaction less than or equal to trace since {:?}: {:?}",
+                prev, since
+            )));
+        }
+
+        trace.update_allow_compaction(since);
         // Atomically update the meta with both the trace and future changes.
         //
         // TODO: Instead of fully overwriting META each time, this should be
         // more like a compactable log.
-        self.blob.set_meta(self.serialize_meta())
+        if let Err(e) = self.blob.set_meta(self.serialize_meta()) {
+            // Revert in-memory state back to its previous version so that
+            // things are consistent between the durably persisted version
+            // and the in-memory version.
+            let trace = self
```

Comment:
I needed to do this (get the trace again) here for lifetime reasons as otherwise I had overlapping mutable and immutable borrows. Lmk if theres a better way to do this / if this deserves a comment

Reply:
Seems fine to me